<a href="https://colab.research.google.com/github/bhouldridge/repertoire/blob/main/SQL_exercise.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# SQL Project Specifications

The goal of this project is to demonstrate the ability to pull data from an SQL database and to perform exploratory data analysis on that data. 

1. **Pick a dataset that interests you** (or multiple data sets) from the Open Data Sets available from Google BigQuery. 
1. **Come up with questions about your data**
  - What sort of information is in this dataset?
  - How many records are there?
  - Have the number of bitcoin transactions increased year over year?
  - Does New Mexico get more or less rain now than 20 years ago? 
  - How many different countries (states, counties, cities, etc) have records in this data set? 
1.  **Use SQL queries to pull the specific information for your questions.**  Make sure and use AT LEAST 13 of the 15 SQL options listed below.  (You may have to get creative and come up with more questions to ask/answer.)
  - Basic Queries
      -  SELECT  (with * and with column names)
      -  WHERE 
      -  AND
      -  OR
      -  LIKE (with % or _ wildcard)
      -  BETWEEN
      -  LIMIT
  - Sorting and Grouping
      -  ORDER BY
      -  DISTINCT
      -  GROUP BY
  - Aggregates
      -  MAX
      -  MIN
      -  SUM
      -  AVG
      -  COUNT
1. **Make some plots.** Make some cool plots to go with your data. Write SQL queries to get ONLY the information you need for each plot.  (Don't pull ALL the data and then just plot a few columns.)

**EXTRA CREDIT:**  
- Use a query that joins two tables.  


#DataSet

##Imports and Authenticate

In [ ]:
from google.cloud import bigquery
from google.colab import auth
import pandas as pd
import plotly.express as px
import plotly.graph_objects as go

auth.authenticate_user()

##Set Up Client

In [ ]:
project_id = 'project-5-288217'
client = bigquery.Client(project=project_id)

##About Dataset

Internet Archive eBook data includes the complete full text of all Internet Archive books published 1800-1922 and all books in the American Libraries collection for which English-language full text was available using the search “collection:(americana)”. These collections have been processed using the GDELT Global Knowledge Graph and are available in Google BigQuery. More than a billion pages stretching back 215 years have been examined to compile a list of all people, organizations, and other names, fulltext geocoded to render them fully mappable, and more than 4,500 emotions and themes compiled. All of this computed metadata is combined with all available book-level metadata, including title, author, publisher, and subject tags as provided by the contributing libraries.

HathiTrust data includes all English language public domain books 1800-2015. They were provided as part of a special research extract and only public domain volumes are included.

##Set Up Dataset Referance

In [ ]:
dataset_ref = client.dataset('internetarchivebooks', project='gdelt-bq')
dataset = client.get_dataset(dataset_ref)

##Print Tables in Dataset

In [ ]:
tables = list(client.list_tables(dataset))
print([table.table_id for table in tables])

['1800', '1801', '1802', '1803', '1804', '1805', '1806', '1807', '1808', '1809', '1810', '1811', '1812', '1813', '1814', '1815', '1816', '1817', '1818', '1819', '1820', '1821', '1822', '1823', '1824', '1825', '1826', '1827', '1828', '1829', '1830', '1831', '1832', '1833', '1834', '1835', '1836', '1837', '1838', '1839', '1840', '1841', '1842', '1843', '1844', '1845', '1846', '1847', '1848', '1849', '1850', '1851', '1852', '1853', '1854', '1855', '1856', '1857', '1858', '1859', '1860', '1861', '1862', '1863', '1864', '1865', '1866', '1867', '1868', '1869', '1870', '1871', '1872', '1873', '1874', '1875', '1876', '1877', '1878', '1879', '1880', '1881', '1882', '1883', '1884', '1885', '1886', '1887', '1888', '1889', '1890', '1891', '1892', '1893', '1894', '1895', '1896', '1897', '1898', '1899', '1900', '1901', '1902', '1903', '1904', '1905', '1906', '1907', '1908', '1909', '1910', '1911', '1912', '1913', '1914', '1915', '1916', '1917', '1918', '1919', '1920', '1921', '1922', '1923notxt', '1

##Set Up Table Referances

In [ ]:
table_ref_1842 = dataset.table('1842')
table_ref_1942notxt = dataset.table('1942notxt')

##Look at Schema

In [ ]:
table_1842 = client.get_table(table_ref_1842)
table_1842.schema

[SchemaField('GKGRECORDID', 'STRING', 'NULLABLE', None, ()),
 SchemaField('DATE', 'INTEGER', 'NULLABLE', None, ()),
 SchemaField('SourceCollectionIdentifier', 'INTEGER', 'NULLABLE', None, ()),
 SchemaField('SourceCommonName', 'STRING', 'NULLABLE', None, ()),
 SchemaField('DocumentIdentifier', 'STRING', 'NULLABLE', None, ()),
 SchemaField('Counts', 'STRING', 'NULLABLE', None, ()),
 SchemaField('V2Counts', 'STRING', 'NULLABLE', None, ()),
 SchemaField('Themes', 'STRING', 'NULLABLE', None, ()),
 SchemaField('V2Themes', 'STRING', 'NULLABLE', None, ()),
 SchemaField('Locations', 'STRING', 'NULLABLE', None, ()),
 SchemaField('V2Locations', 'STRING', 'NULLABLE', None, ()),
 SchemaField('Persons', 'STRING', 'NULLABLE', None, ()),
 SchemaField('V2Persons', 'STRING', 'NULLABLE', None, ()),
 SchemaField('Organizations', 'STRING', 'NULLABLE', None, ()),
 SchemaField('V2Organizations', 'STRING', 'NULLABLE', None, ()),
 SchemaField('V2Tone', 'STRING', 'NULLABLE', None, ()),
 SchemaField('Dates', 'ST

In [ ]:
table_1942notxt = client.get_table(table_ref_1942notxt)
table_1942notxt.schema

[SchemaField('GKGRECORDID', 'STRING', 'NULLABLE', None, ()),
 SchemaField('DATE', 'INTEGER', 'NULLABLE', None, ()),
 SchemaField('SourceCollectionIdentifier', 'INTEGER', 'NULLABLE', None, ()),
 SchemaField('SourceCommonName', 'STRING', 'NULLABLE', None, ()),
 SchemaField('DocumentIdentifier', 'STRING', 'NULLABLE', None, ()),
 SchemaField('Counts', 'STRING', 'NULLABLE', None, ()),
 SchemaField('V2Counts', 'STRING', 'NULLABLE', None, ()),
 SchemaField('Themes', 'STRING', 'NULLABLE', None, ()),
 SchemaField('V2Themes', 'STRING', 'NULLABLE', None, ()),
 SchemaField('Locations', 'STRING', 'NULLABLE', None, ()),
 SchemaField('V2Locations', 'STRING', 'NULLABLE', None, ()),
 SchemaField('Persons', 'STRING', 'NULLABLE', None, ()),
 SchemaField('V2Persons', 'STRING', 'NULLABLE', None, ()),
 SchemaField('Organizations', 'STRING', 'NULLABLE', None, ()),
 SchemaField('V2Organizations', 'STRING', 'NULLABLE', None, ()),
 SchemaField('V2Tone', 'STRING', 'NULLABLE', None, ()),
 SchemaField('Dates', 'ST

##Are 1842 and 1942 Schema Different?

In [ ]:
str(table_1942notxt.schema) == str(table_1842.schema)

False

Not the same schema but I'm not going to dig through to find out where.

##Look at Some Entries

In [ ]:
query = """
        SELECT *
        FROM `gdelt-bq.internetarchivebooks.1842`
        LIMIT 5
        """
df = client.query(query).to_dataframe()
df.head(5)

GKGRECORDID  ...                                  BookMeta_FullText
0  iabook-chapelhymnsselec00maso_0  ...   CHAPEL HYMNS: f SELECTION OF HYMNS. m «TTH ^f...
1                  iabook-b1502917  ...   THE LIBRARY OF THE METROPOLITAN MUSEUM OF ART...
2            iabook-ASPC0001938500  ...   AN APPEAL TO THE PROFESSORS OF CHRISTIANITY, ...
3            iabook-ASPC0001981600  ...   1)11. CllAJiNlNG'S LA.ST ADDlil^SB,'^ DEUVEiU...
4            iabook-ASPC0002432300  ...   '1 I 1* ^ 4 It S«a CAl.LENDER COLLECTION 'niE...

[5 rows x 47 columns]

Many of the columns such as V2Tone and Datees don't make intuitive sence, but colums like BookMeta_ScannedImages, and BookMeta_Added Date will be useful.

#Questions & Queries

##Setup Safe Query

In [ ]:
GB = 10 ** 9

safe_config = bigquery.QueryJobConfig(maximum_bytes_billed=GB)

##How many rows?

In [ ]:
query = """
        SELECT COUNT(*) AS total_books
        FROM `gdelt-bq.internetarchivebooks.1842`
        """
row_count = client.query(query).to_dataframe()
row_count

total_books
0         3133

3133 books are found in this table.

##Most downloads for a book?

In [ ]:
query = """
        SELECT MAX(BookMeta_DownloadsJune2015) AS max_downloads
        FROM `gdelt-bq.internetarchivebooks.1842`
        """
df = client.query(query).to_dataframe()
print(df)

   max_downloads
0           3079


It looks like one of the books has been downloaded 3079 times!

##Which book has the most downloads?

In [ ]:
query = """
        SELECT BookMeta_Title AS popular_book
        FROM `gdelt-bq.internetarchivebooks.1842`
        WHERE BookMeta_DownloadsJune2015 = 3079
        """
df = client.query(query).to_dataframe()
print(df)

                                        popular_book
0  The banquet of Dun na n-Gedh and The battle of...


"The Banquet of Dun Na N-Gedh and the Battle of Magh Rath" is the most downloaded book from 1842. 

##Which book has the least downloads?

In [ ]:
query = """
        SELECT MIN(BookMeta_DownloadsJune2015) AS min_downloads
        FROM `gdelt-bq.internetarchivebooks.1842`
        """
df = client.query(query).to_dataframe()
print(df)

   min_downloads
0              1


In [ ]:
query = """
        SELECT BookMeta_Title AS forgetten_books
        FROM `gdelt-bq.internetarchivebooks.1842`
        WHERE BookMeta_DownloadsJune2015 = 1
        """
df = client.query(query).to_dataframe()
print(df)

                                     forgetten_books
0                    The Baptist missionary magazine
1           Lowell Cemetery deeds by year, 1842-1864
2  The lives of the fathers, martyrs, and other p...
3  The lives of the fathers, martyrs, and other p...
4                    The Baptist missionary magazine


Five of the books in this table have been downloaded only once. All the books have been downloaded at least once.

##What is the average downloads for a book?

In [ ]:
query = """
        SELECT AVG(BookMeta_DownloadsJune2015) AS avg_downloads
        FROM `gdelt-bq.internetarchivebooks.1842`
        """
df = client.query(query).to_dataframe()
print(df)

   avg_downloads
0     279.865341


The average book is downloaded about 280 times.

##Which books have been downloaded between 250 and 300 times?

In [ ]:
query = """
        SELECT BookMeta_Title AS Books, BookMeta_DownloadsJune2015 AS downloads
        FROM `gdelt-bq.internetarchivebooks.1842`
        WHERE BookMeta_DownloadsJune2015 BETWEEN 250 AND 300
        ORDER BY BookMeta_DownloadsJune2015
        """
df = client.query(query).to_dataframe()
print(df)

                                                 Books  downloads
0         The nabob at home; or, The return to England        250
1    A treatise on the incarnation of the Eternal Word        250
2    A Dictionary of Two Thousand Italian, French, ...        250
3              Moral agency : and man as a moral agent        250
4    An address on the early settlement of the vall...        250
..                                                 ...        ...
233  Wealth and pedigree of the wealthy citizens of...        299
234  The life of George Washington : written for th...        299
235  Observations on the salmon fisheries of Ulster...        299
236  View of Ancient and Modern Egypt: With an Outl...        300
237  An essay on the means and importance of introd...        300

[238 rows x 2 columns]


238 of the books have been downloaded between 250 and 300 times.


##What values are in TranslationInfo?

In [ ]:
query = """
        SELECT DISTINCT TranslationInfo
        FROM `gdelt-bq.internetarchivebooks.1842`
        """
df = client.query(query, job_config=safe_config).to_dataframe()
print(df)

  TranslationInfo
0            None


There is no Translation Info for 1842.

##What is the total number of scanned images for 1842?

In [ ]:
query = """
        SELECT SUM(BookMeta_ScannedImages) AS total_images
        FROM `gdelt-bq.internetarchivebooks.1842`
        """
client.query(query).to_dataframe()

total_images
0        911002

The number of images in all the books is 911,002

##Who are the most prolific creators in 1842?

In [ ]:
query = """
        SELECT  BookMeta_Creator AS creator, COUNT(1) AS books
        FROM `gdelt-bq.internetarchivebooks.1842`
        GROUP BY BookMeta_Creator
        ORDER BY books DESC
        """
df = client.query(query).to_dataframe()
df.head(5)

creator  books
0                                               None    146
1  New York (N.Y.). Common Council;Willis, Samuel...     16
2                      Beveridge, William, 1637-1708     15
3    Gore, Mrs. (Catherine Grace Frances), 1799-1861     12
4  American Baptist Missionary Union. Executive C...     12

There are many books without a creator. Other Than that New York Common Council and Beveridge, William are the most common.

##What books did they contriute?

In [ ]:
query = """
        SELECT BookMeta_Creator AS creator, BookMeta_Title AS book
        FROM `gdelt-bq.internetarchivebooks.1842`
        WHERE BookMeta_Creator = 'Beveridge, William, 1637-1708' OR
              BookMeta_Creator LIKE 'New York (N.Y.). Common Council;Willis, Samuel%'
              ORDER BY creator
        """
df = client.query(query).to_dataframe()
print(df)

                                              creator                                               book
0                       Beveridge, William, 1637-1708         The theological works of William Beveridge
1                       Beveridge, William, 1637-1708  The theological works of William Beveridge, D....
2                       Beveridge, William, 1637-1708  The theological works of William Beveridge, D....
3                       Beveridge, William, 1637-1708         The theological works of William Beveridge
4                       Beveridge, William, 1637-1708  The theological works of William Beveridge, D....
5                       Beveridge, William, 1637-1708  The theological works of William Beveridge, D....
6                       Beveridge, William, 1637-1708  The theological works of William Beveridge, D....
7                       Beveridge, William, 1637-1708  The theological works of William Beveridge, D....
8                       Beveridge, William, 1637-1708  

Looks like the Common Council of New York produced Manuals and William Beveridge has several editions of his Theological works represented. 

##When were books added to the dataset? 

In [ ]:
query = """
        SELECT  BookMeta_AddedDate, COUNT(1) AS number_of_books
        FROM `gdelt-bq.internetarchivebooks.1842`
        GROUP BY BookMeta_AddedDate
        ORDER BY number_of_books DESC
        """
df = client.query(query).to_dataframe()
print(df)

    BookMeta_AddedDate  number_of_books
0                 None             1345
1           2009-10-07               39
2           2014-07-15               19
3           2012-09-27               13
4           2015-05-14               12
..                 ...              ...
840         2012-03-22                1
841         2012-12-11                1
842         2012-09-10                1
843         2010-01-05                1
844         2009-03-19                1

[845 rows x 2 columns]


Looks like the date added was not recorded for most of the books.

#Plots

##Plot a histogram of how many times the books have been downloaded.  

Create data frame for graph.

In [ ]:
query = """
        SELECT BookMeta_DownloadsJune2015 AS downloads
        FROM `gdelt-bq.internetarchivebooks.1842`
        """
df_downloads = client.query(query).to_dataframe()
print(df_downloads)

      downloads
0          12.0
1         204.0
2         204.0
3         133.0
4         111.0
...         ...
3128      212.0
3129      128.0
3130      103.0
3131      131.0
3132       27.0

[3133 rows x 1 columns]


Plot downloads

In [ ]:
fig = px.histogram(df_downloads,
                   x = 'downloads',
                   title= 'Distribution of Downloads',
                   nbins=1000)
fig.show()

Histogram shows two distinct peaks, one close to zero and another around 200.  Very few books have been downloaded more than 1000 times.

##Plot a histogram of how many images are in each book.

Create data frame for graph.

In [ ]:
query = """
        SELECT BookMeta_ScannedImages	 AS Images
        FROM `gdelt-bq.internetarchivebooks.1842`
        """
df_images = client.query(query).to_dataframe()
print(df_images)

      Images
0       67.0
1       26.0
2      310.0
3       70.0
4      565.0
...      ...
3128   659.0
3129   193.0
3130   305.0
3131    96.0
3132   544.0

[3133 rows x 1 columns]


Plot images

In [ ]:
fig = px.histogram(df_images,
                   x = 'Images',
                   title= 'Distribution of Images per Creator',
                   nbins=300)
fig.show()

Again two distinct peaks, one close to zero and one close to 325 images. Few books have more than 800 images.

##Images and Downloads Plot

Produce a data frame that contains each creator with the number of downloads and images attributed to them. Filter out the none data by caping images at 12,000. 

In [ ]:
query = """
        SELECT BookMeta_Creator AS Creators,
        SUM(BookMeta_ScannedImages) AS images,
        SUM(BookMeta_DownloadsJune2015) AS downloads
        FROM `gdelt-bq.internetarchivebooks.1842`
        GROUP BY Creators
        HAVING images < 12000
        ORDER BY images
        """
df_creators = client.query(query).to_dataframe()
print(df_creators)

                                               Creators  images  downloads
0     Knight, Anne;Chapman, Maria Weston, 1806-1885,...       2       82.0
1     Chapman, Maria Weston, 1806-1885;Borden, Natha...       4       99.0
2     Smith, Increase S;Weston, Caroline, 1808-1882,...       4       72.0
3     Cook, Russell S., 1811-1864;Phelps, Amos A. (A...       4        NaN
4     Chapman, Maria Weston, 1806-1885;Child, David ...       4       93.0
...                                                 ...     ...        ...
1996      Bale,  John ,  Coverdale,  Miles ,  John Foxe    4904     1860.0
1997  Butler, Alban, 1711-1773;Butler, Charles, 1750...    5502       45.0
1998  Emmons, Nathanael, 1745-1840;Ide, Jacob, 1785-...    5786     2599.0
1999                      Beveridge, William, 1637-1708    7832     1429.0
2000  New York (N.Y.). Common Council;Willis, Samuel...   11406    25694.0

[2001 rows x 3 columns]


Plot the number of images and downloads for each creator. 

In [ ]:
fig = px.scatter(df_creators,
                x = 'images',
                y = 'downloads',
               hover_name = 'Creators')
fig.update_layout(title='Creators by Images and Downloads')
fig.show()

Most of creators have between 0 and 5000 downloads and 1 to 2000 images. Just one overachever, New York (N.Y.). Common Council push the  maximum numbers quite high. 